keylime/allowlistdbproxy.py

```diff
@@ -0,0 +1,177 @@
+import threading
+import time
+
+from sqlalchemy import select
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.orm.exc import NoResultFound
+
+from keylime import keylime_logging
+from keylime.db.verifier_db import VerifierAllowlist
+
+logger = keylime_logging.init_logging("allowlistdbproxy")
+
+
+class AllowlistEntry:
```

Comment:
I like this approach. Did you research a more generic approach? `functools` has `lru_cache` that it is not supporting `ttl`, but I think that should be fairly easy to add.

Having some kind generic reusable memoization class can help also in different scenarios in keylime.  This class can be used as a decorator for specific methods, and later can be added persistence too (one example can be [2])

[1] https://docs.python.org/3/library/functools.html#functools.lru_cache
[2] https://github.com/openSUSE/openSUSE-release-tools/blob/master/osclib/memoize.py

Reply:
The `lru_cache` may be a bit older than the `cache` python version-wise.

From what I see we need to constantly poll the DB for an update of AllowlistVerifier table so we don't really have a constant cache but have to explicitly go out to the DB every time (or every 0.1s of a non-update) .

```
            # We have to be really careful with this update_timeout. If attestation fails
            # and an update of the allowlist comes very quickly this update_timeout may
            # cause the new policy not to be loaded and cause an unexpected attestation
            # failure... it may be better to always poll the DB.
```


```
 @functools.lru_cache(user_function)
 @functools.lru_cache(maxsize=128, typed=False)

    Decorator to wrap a function with a memoizing callable that saves up to the maxsize most recent calls. It can save time when an expensive or I/O bound function is periodically called with the same arguments.
```

lease-recently used doesn't sound right. I would need something that times out unused elements but not something like a limited-size LRU queue.

Reply:
> The `lru_cache` may be a bit older than the `cache` python version-wise.

Right. `cache` is expressed as a `lru_cache` with different defaults, so can be more handy. IIRC is just a two lines decorator.
 
> lease-recently used doesn't sound right. I would need something that times out unused elements but not something like a limited-size LRU queue.

Yes. That was I was referring as a "different invalidation than a TTL". As commented before, a time-to-live parameter should be fairly easy to implement on top of `lru_cache`.

The second link of my first comment is a different decorator (not using `lru_cache`, as the original code comes from 2.7) that it is implementing a TTL invalidator.

In any case, I think that we can move on : ))